File: spark_java_study/src/main/java/com/zhangzl/study/sql/RDD2DataFrameReflection.java
package com.zhangzl.study.sql;

import com.zhangzl.study.domain.Student;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;

import java.util.List;

/**
 * Description  : 通过反射机智，使得RDD转换为dataFrame
 * Class   Path : com.zhangzl.study.sql
 * Create  User : zhangzuolong
 * Create  Time : 2017/3/24 17:03
 * Project Name : technology_sophisticated_study
 *
 * Java版本：Spark SQL是支持将包含了JavaBean的RDD转换为DataFrame的。
 * JavaBean的信息，就定义了元数据。
 * Spark SQL现在是不支持将包含了嵌套JavaBean或者List等复杂数据的JavaBean，
 * 作为元数据的。只支持一个包含简单数据类型的field的JavaBean。
 *
 */
public class RDD2DataFrameReflection {

    public static void main(String[] args) {

        SparkConf  conf = new SparkConf().setMaster("local").setAppName("rdd2DataFrameReflaction");
        JavaSparkContext sc = new JavaSparkContext(conf);
        SQLContext sqlContext = new SQLContext(sc);

        JavaRDD<String> students = sc.textFile("D:\\workspaces\\idea\\technology_sophisticated_study\\spark_java_study\\src\\main\\resources\\conf\\students.txt");


        JavaRDD<Student> studentsRdd = students.map(new Function<String, Student>() {
            public Student call(String line) throws Exception {
                String[]   lineSplits= line.split(",");
                Student student = new Student();
                student.setId(Integer.valueOf(lineSplits[0].trim()));
                student.setName(lineSplits[1]);
                student.setAge(Integer.valueOf(lineSplits[2].trim()));
                return student;
            }
        });

        /**
         * 使用反射方式，将RDD转换为DataFrame
         // 将Student.class传入进去，其实就是用反射的方式来创建DataFrame
         // 因为Student.class本身就是反射的一个应用
         // 然后底层还得通过对Student Class进行反射，来获取其中的field
         // 这里要求，JavaBean必须实现Serializable接口，是可序列化的
         */
        DataFrame  studentDataFrame = sqlContext.createDataFrame(studentsRdd,Student.class);

        //利用dataFrame 建立一个临时表
        studentDataFrame.registerTempTable("student");
        //利用万能的sql来过滤数据
        DataFrame teenegersDataFrame = sqlContext.sql("select * from student where age >=17 ");
        //将DataFrame 在转换为rdd
        JavaRDD<Row> teenegerRdd = teenegersDataFrame.toJavaRDD();

        JavaRDD<Student> teenegers = teenegerRdd.map(new Function<Row, Student>() {
            public Student call(Row row) throws Exception {

                Student student = new Student();
                student.setAge(Integer.valueOf(row.getAs("age").toString()));
                student.setName(row.getAs("name").toString());
                student.setId(Integer.valueOf(row.getAs("id").toString()));
                return  student;
            }
        });

        List<Student> list = teenegers.collect();
        for(Student s : list){
            System.out.println(s.toString());
        }

    }
}

File: spark_scala_study/src/main/scala/com/zhangzl/study/sql/RDD2DataFrameReflection.scala
package com.zhangzl.study.sql

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.sql.SQLContext


/**
  * Description  : RDD转为DataFrame
  * Class   Path : com.zhangzl.study.sql
  * Create  User : zhangzuolong
  * Create  Time : 2017/3/24 21:10
  * Project Name : technology_sophisticated_study
  *
  * Scala版本：而Scala由于其具有隐式转换的特性，所以Spark SQL的Scala接口，
  * 是支持自动将包含了case class的RDD转换为DataFrame的。
  * case class就定义了元数据。Spark SQL会通过反射读取传递给case class的参数的名称，
  * 然后将其作为列名。与Java不同的是，Spark SQL是支持将包含了嵌套数据结构的case class作为元数据的，
  * 比如包含了Array等
  */
object RDD2DataFrameReflection extends App{

    //建立基本属性
    val conf =  new SparkConf().setMaster("local").setAppName("rdd2DataFrame")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)

    // 在Scala中使用反射方式，进行RDD到DataFrame的转换，需要手动导入一个隐式转换
    import sqlContext.implicits._

    case class Student(id:Int,name:String,age:Int)

    //
    val studentDF = sc.textFile("D:\\workspaces\\idea\\technology_sophisticated_study\\spark_scala_study\\src\\main\\resources\\conf\\students.txt",1)
                      .map{line => line.split(",")} //按逗号切割
                      .map{arrys => Student(arrys(0).trim.toInt,arrys(1),arrys(2).trim.toInt)}
                      .toDF()  //转换为dataframe
    studentDF.registerTempTable("student")

    val teenagerDF = sqlContext.sql("select * from student where age>17")
    teenagerDF.rdd
      .map{row => Student(row.getAs("id").toString.toInt,row.getAs("name").toString,row.getAs("age").toString.toInt)}
      .foreach{student => println(student.id+":"+student.name+"："+student.age)}
















}
